modules/ray_tracing/src/utils/RayTracingAngularQuadrature.C

```diff
@@ -0,0 +1,276 @@
+    {
+      // If 2D, throw away the downward mu
+      if (_dim == 2 && gauss_legendre_x[j] < 0.5 - TOLERANCE * TOLERANCE)
+        continue;
+
+      _phi.push_back(chebyshev_x[i]);
+      _mu.push_back(gauss_legendre_x[j] * (_mu_max - _mu_min) + _mu_min);
+      _polar_sin.push_back(std::sin(std::acos(_mu.back())));
+
+      const Real weight_factor =
+          (_dim == 2 && !MooseUtils::absoluteFuzzyEqual(gauss_legendre_x[j], 0.5)) ? 2.0 : 1.0;
+      _w.push_back(weight_factor * chebyshev_w[i] * gauss_legendre_w[j]);
+    }
+}
+
+void
+RayTracingAngularQuadrature::gaussLegendre(const unsigned int order,
+                                           std::vector<Real> & x,
+                                           std::vector<Real> & w)
+{
+  if (order == 0)
+    mooseError("Order must be positive in gaussLegendre()");
+
+  x.resize(order);
+  w.resize(order);
+  libMesh::DenseMatrix<Real> mat(order, order);
+  libMesh::DenseVector<Real> lambda(order);
+  libMesh::DenseVector<Real> lambdai(order);
+  libMesh::DenseMatrix<Real> vec(order, order);
+
+  for (unsigned int i = 1; i < order; ++i)
+  {
+    Real ri = i;
+    mat(i, i - 1) = ri / std::sqrt(((2. * ri - 1.) * (2. * ri + 1.)));
+    mat(i - 1, i) = mat(i, i - 1);
+  }
+  mat.evd_right(lambda, lambdai, vec);
+
+  for (unsigned int i = 0; i < order; ++i)
+  {
+    x[i] = 0.5 * (lambda(i) + 1.0);
+    w[i] = vec(0, i) * vec(0, i);
+  }
+
+  // Sort based on the points
+  std::vector<std::size_t> sorted_indices(x.size());
+  std::iota(sorted_indices.begin(), sorted_indices.end(), 0);
+  std::stable_sort(sorted_indices.begin(), sorted_indices.end(), [&x](size_t i1, size_t i2) {
+    return x[i1] < x[i2];
+  });
+  const auto x_copy = x;
+  const auto w_copy = w;
+  for (std::size_t i = 0; i < x.size(); ++i)
+  {
+    x[i] = x_copy[sorted_indices[i]];
+    w[i] = w_copy[sorted_indices[i]];
+  }
+}
+
+void
+RayTracingAngularQuadrature::rotate(const libMesh::Point & rotation_direction)
+{
+  _current_rotation_direction = rotation_direction;
+
+  libMesh::DenseMatrix<Real> rotation_matrix;
+  rotationMatrix(rotation_direction.unit(), rotation_matrix);
+
+  _current_directions.clear();
+  _current_directions.reserve(_phi.size());
+
+  _current_weights.clear();
+  _current_weights.reserve(_w.size());
+
+  _current_polar_sins.clear();
+  _current_polar_sins.reserve(_polar_sin.size());
+
+  libMesh::DenseVector<Real> omega(3);
+  libMesh::DenseVector<Real> omega_p(3);
+  Point direction;
+
+  for (std::size_t q = 0; q < _phi.size(); ++q)
+  {
+    // Get direction as a DenseVector for rotation
+    omega(0) = sqrt(1 - _mu[q] * _mu[q]) * cos(_phi[q]);
+    omega(1) = sqrt(1 - _mu[q] * _mu[q]) * sin(_phi[q]);
+    omega(2) = _mu[q];
+
+    // Rotate and create the direction
+    rotation_matrix.vector_mult(omega_p, omega);
+    direction(0) = omega_p(0);
+    direction(1) = omega_p(1);
+    direction(2) = omega_p(2);
+
+    std::size_t l = _current_directions.size();
```

Comment:
everything from here to line 182 is really confusing code mainly because l is set and then the value of l is checked later. Can this be simplified?

Reply:
I can make the code a bit longer but easier to read.